src/grlib.c

```diff
@@ -1171,109 +1026,68 @@ apbuart5_init (void)
   uart_init(&uarts[5]);
 }
 
+static void
+uart_rx_event (int32 arg)
+{
+  apbuart_type *uart = get_uart_by_irq (arg);
+
+  if (uart != NULL)
+  {
+    if (apbuart_read_event (uart) && apbuart_get_flag (uart->control_register, APBUART_RI))
+    {
+      grlib_set_irq (uart->irq);
+    }
+
+    event (uart_rx_event, uart->irq, UART_RX_TIME);
+  }
+}
+
+static void
+fast_uart_rx_event (int32 arg)
+{
+  apbuart_type *uart = get_uart_by_irq (arg);
+
+  if (uart != NULL)
+  {
+    if (apbuart_fast_read_event (uart) && apbuart_get_flag (uart->control_register, APBUART_RI))
+    {
+      grlib_set_irq (uart->irq);
+    }
+
+    event (fast_uart_rx_event, uart->irq, UART_RX_TIME);
+  }
+}
+
 int
 uart_read (apbuart_type *uart, uint32_t addr, uint32_t *data)
 {
   int result = 1;
-  unsigned tmp = 0;
 
   switch (addr & APBUART_REGISTER_TYPE_MASK)
   {
     case APBUART_DATA_REGISTER_ADDRESS:
-#ifndef _WIN32
+      apbuart_reset_flag(&uart->status_register, APBUART_DR);
+      *data = uart->uart_io.in.buffer[uart->uart_io.in.buffer_index];
 #ifdef FAST_UART
-      if (uart->in_stream.buffer_index < uart->in_stream.buffer_size)
-      {
-        if ((uart->in_stream.buffer_index + 1) < uart->in_stream.buffer_size)
-        {
-          grlib_set_irq (uart->irq);
-        }
-        *data = (uint32) uart->in_stream.buffer[uart->in_stream.buffer_index++];
-        result = 0;
-      }
-      else
-      {
-        if (uart->device_open)
-        {
-          uart->in_stream.buffer_size = DO_IO_READ (uart->in_stream.descriptor, uart->in_stream.buffer, APBUART_BUFFER_SIZE);
-        }
-        else
-        {
-          uart->in_stream.buffer_size = 0;
-        }
-
-        if (uart->in_stream.buffer_size > 0)
-        {
-          uart->in_stream.buffer_index = 0;
-          if ((uart->in_stream.buffer_index + 1) < uart->in_stream.buffer_size)
-          {
-            grlib_set_irq (uart->irq);
-          }
-          *data = (uint32) uart->in_stream.buffer[uart->in_stream.buffer_index++];
-        }
-        else
-        {
-          *data = (uint32) uart->in_stream.buffer[uart->in_stream.buffer_index];
-        }
-        
-        result = 0;
-      }
-#else
-      uart->status_register &= ~APBUART_STATUS_REG_DATA_READY;
-      *data = (uint32) uart->in_stream.data;
-      result = 0;
+      fast_uart_rx_event (uart->irq);
```

Comment:
That line exist for immediate read new data in file in case of FAST UART. If FAST UART is not set, a read event is triggered periodically by event loop regardless of the data being read by the user. This immediate read is legacy functionality.